04-onion-routing.md

```diff
@@ -471,37 +476,38 @@ The processing node:
     - MUST drop the packet.
     - MUST signal a route failure.
 
+[FIXME: separate processing node requirements into `Requirements` section, or is more it important to integrate them into the sequence of events?]
+
 # Shared Secret
 
 The origin node performs ECDH with each hop of the route, in order to establish a secret.
 A new _sessionkey_, a 32-byte EC private key, is generated for each message.
 The shared secret builder function takes a public key and a 32-byte secret
-as input and returns a 32-byte secret as output.
+,as inputs, and returns a 32-byte secret, as output.
 
 During the packet generation phase, the secret is the `sessionkey`, and the
-public key is the processing node's public key, which is
-blinded by all previous blinding factors.
-During the processing phase, the secret is the processing node's private key,
-while the public key is the ephemeral public key from the packet, which has been
-incrementally blinded by its predecessors.
+public key is the hop's public key, which is blinded by all previous blinding
+factors.
+During the processing phase, the secret is the hop's (processing node's) private
+key, while the public key is the ephemeral public key from the packet, which has
+been incrementally blinded by its predecessors.
 
 The public key is multiplied by the secret, using the `secp256k1` curve.
 The DER compressed representation of the multiplication result is serialized and
-hashed using `SHA256`.
-The resulting hash is returned as the shared secret.
+hashed using `SHA256`, and the resulting hash is returned as the shared secret.
 Notice that this is the ECDH variant implemented in `libsecp256k1`.
 
 # Filler Generation
 
 Upon receiving a packet, the processing node extracts the information destined
 for it from the route information and the per-hop payload.
 The extraction is done by deobfuscating and left-shifting the field.
-This would make the field shorter at each hop, allowing an attacker to deduce the route length.
-For this reason, the field is padded before forwarding.
+This would make the field shorter at each hop, allowing an attacker to deduce the
+route length. For this reason, the field is pre-padded before forwarding.
 Since the padding is part of the HMAC, the origin node will have to generate an
-identical padding in order to compute the HMACs correctly for each hop.
+identical [FIXME: identical to what?] padding in order to compute the HMACs correctly for each hop.
```

Comment:
The padding generated by the origin node when it creates the packet must be the same as that produced by the processing node when it pads the packet.  Otherwise, the next hop will fail because the HMAC will be wrong...

Reply:
gotcha, thanks

Reply:
...pre-generate an identical padding (to that which each hop will generate)...